### ISSUE TITLE
Performance issue with FILTER statement

### ISSUE THREAD

**Issue by mathbagu** [eclipse/rdf4j#2408]
We are updating from an old version of RDF (2.4.4) to a recent one, but we are facing to a performance issue for one of our queries:
```
SELECT *
WHERE {
{ GRAPH ?graph {
    ?Switch
        a ?type ;
        cim:IdentifiedObject.name ?name ;
        cim:Equipment.EquipmentContainer ?EquipmentContainer .
    VALUES ?type { cim:Switch cim:Breaker cim:Disconnector cim:LoadBreakSwitch cim:ProtectedSwitch } .
    ?Terminal1
        a cim:Terminal ;
        cim:Terminal.ConductingEquipment ?Switch .
    ?Terminal2
        a cim:Terminal ;
        cim:Terminal.ConductingEquipment ?Switch .
    FILTER ( STR(?Terminal1) < STR(?Terminal2) )
}}
OPTIONAL { GRAPH ?graphSSH {
    ?Switch cim:Switch.open ?open
}}
}
```
On our dataset, the test case takes:
- 3-4 seconds with v2.4.4 to v3.1.4
- 1m30 with v3.2.0
- 1m with v3.3.0

Without the `FILTER` statement, the performance are OK (comparable to old versions).

**Comment by hmottestad** [comment]
Could you take a look at the query plan with: https://rdf4j.org/documentation/programming/repository/#explaining-queries

**Comment by hmottestad** [comment]
This may be related to scoping issues that were fixed at sometime in 3.1.x. 

Maybe moving the line with "VALUES" up to be right above "?Switch" would help...??

**Comment by mathbagu** [comment]
Thanks to the documentation, and the `explanation` feature, I fixed my queries and the performance are quite similar to what I had with the previous release.

For your information, just moving the `VALUES` wasn't enough. In my query, a new scope was created with FILTER that takes a lot of time because of the number of Terminals in my dataset.
